### PR TITLE
Fix for too many nonfinalized blocks at the start of the node

### PR DESCRIPTION
**Pull request by kostekIV**

### CODE REVIEW

**Review by timorleph — APPROVED** [review]
Lovely. Maybe add a comment about this likely not being needed after A0-2984?

**Comment by kostekIV** [comment]
> Lovely. Maybe add a comment about this likely not being needed after A0-2984?

added

File: bin/node/src/service.rs [hunk 1]
```diff
@@ -64,6 +64,21 @@ impl<N: BaseArithmetic> BackoffAuthoringBlocksStrategy<N> for LimitNonfinalized
...
+fn revert_unfinalized_blocks(client: &FullClient) {
+    let last_finalized = client.info().finalized_number;
+    let best_block = client.info().best_number;
+    // 20 so there are no problems with finality session
+    const REVERT_THRESHOLD: u32 = 20;
+
+    let blocks_count_to_revert = best_block.saturating_sub(last_finalized + REVERT_THRESHOLD);
+
+    debug!("Reverting {} not finalized blocks", blocks_count_to_revert);
+    let new_best = client
```
Comment (lesniak43) [inline bin/node/src/service.rs:77]: I think it returns the number of blocks that were reverted, not the new best block number.

**Review by lesniak43 — APPROVED** [review]

File: bin/node/src/service.rs [hunk 2]
```diff
@@ -64,6 +64,22 @@ impl<N: BaseArithmetic> BackoffAuthoringBlocksStrategy<N> for LimitNonfinalized
...
+/// This will not be needed after A0-2984 is done
+fn revert_unfinalized_blocks(client: &FullClient) {
+    let last_finalized = client.info().finalized_number;
+    let best_block = client.info().best_number;
+    // 20 so there are no problems with finality session
+    const REVERT_THRESHOLD: u32 = 20;
+
+    let blocks_count_to_revert = best_block.saturating_sub(last_finalized + REVERT_THRESHOLD);
+
+    debug!("Reverting {} not finalized blocks", blocks_count_to_revert);
```
Comment (DamianStraszak) [inline bin/node/src/service.rs:76]: Could you make this `info`? And also add `last_finalized`, `best_block` to the log message?

File: bin/node/src/service.rs [hunk 3]
```diff
@@ -64,6 +64,22 @@ impl<N: BaseArithmetic> BackoffAuthoringBlocksStrategy<N> for LimitNonfinalized
...
+    let new_best = client
+        .revert(blocks_count_to_revert)
+        .expect("Error while reverting blocks: ");
+    debug!("Reverted chain to block #{}", new_best);
```
Comment (DamianStraszak) [inline bin/node/src/service.rs:80]: Same here, please make this `info`

File: bin/node/src/service.rs (hunk 2 above)
Comment (DamianStraszak) [reply bin/node/src/service.rs:76]: Also, let's wrap this into `if blocks_count_to_revert > 0`

File: bin/node/src/service.rs (hunk 1 above)
Comment (kostekIV) [reply bin/node/src/service.rs]: good catch, fixed

File: bin/node/src/service.rs (hunk 2 above)
Comment (kostekIV) [reply bin/node/src/service.rs]: done

File: bin/node/src/service.rs (hunk 3 above)
Comment (kostekIV) [reply bin/node/src/service.rs]: done

**Pull request closed without merging**